### ISSUE TITLE
CRDs for the Operator are generated multiple times during the build

### ISSUE THREAD

**Issue by baloo42** [keycloak/keycloak#32554]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

operator

### Describe the bug

The CRDs for the Operator are generated multiple times during the build.

### Version

main

### Regression

- [ ] The issue is a regression

### Expected behavior

- The CRDs are generated only once. 
- Only one strategy is used (either quarkus-operator-sdk or crd-generator-apt)

### Actual behavior

The CRDs are generated two times, first by crd-generator-apt then by quarkus-operator-sdk/quarkus-maven-plugin.

### How to Reproduce?

```
cd operator
mvn clean package
```

### Anything else?

Logs:
```
[INFO] Scanning for projects...
[INFO] 
[INFO] -------------------< org.keycloak:keycloak-operator >-------------------
[INFO] Building Keycloak Operator 999.0.0-SNAPSHOT
[INFO]   from pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- enforcer:3.0.0-M3:enforce (enforce-java-version) @ keycloak-operator ---
[INFO] 
[INFO] --- enforcer:3.0.0-M3:enforce (enforce-maven-version) @ keycloak-operator ---
[INFO] 
[INFO] --- resources:3.2.0:copy-resources (copy-resources) @ keycloak-operator ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Using 'UTF-8' encoding to copy filtered properties files.
[INFO] Copying 3 resources
[INFO] 
[INFO] --- resources:3.2.0:copy-resources (copy-ubi-null) @ keycloak-operator ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Using 'UTF-8' encoding to copy filtered properties files.
[INFO] Copying 1 resource
[INFO] 
[INFO] --- buildnumber:1.4:create (get-scm-revision) @ keycloak-operator ---
[INFO] Executing: /bin/sh -c cd '/home/bernhard/git/github/keycloak/keycloak/operator' && 'git' 'rev-parse' '--verify' 'HEAD'
[INFO] Working directory: /home/bernhard/git/github/keycloak/keycloak/operator
[INFO] Storing buildNumber: 2ab9d04136a6d8112c499ab80a8ffbcfba618507 at timestamp: 1725205957844
[INFO] Storing buildScmBranch: main
[INFO] 
[INFO] --- build-helper:3.2.0:add-source (default) @ keycloak-operator ---
[INFO] Source directory: /home/bernhard/git/github/keycloak/keycloak/operator/target/generated-sources/java added.
[INFO] 
[INFO] --- resources:3.2.0:resources (default-resources) @ keycloak-operator ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Using 'UTF-8' encoding to copy filtered properties files.
[INFO] Copying 3 resources
[INFO] 
[INFO] --- compiler:3.8.1:compile (default-compile) @ keycloak-operator ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 43 source files to /home/bernhard/git/github/keycloak/keycloak/operator/target/classes
 96% Generating: org.keycloak.operator.crds.v2alpha1.deployment.Keycloak[INFO] 100%: Builder generation complete.                                                                                      
[INFO] Generating CRD keycloakrealmimports.k8s.keycloak.org:
[INFO]   - v1beta1 -> /home/bernhard/git/github/keycloak/keycloak/operator/target/classes/META-INF/fabric8/keycloakrealmimports.k8s.keycloak.org-v1beta1.yml
[INFO]   - v1 -> /home/bernhard/git/github/keycloak/keycloak/operator/target/classes/META-INF/fabric8/keycloakrealmimports.k8s.keycloak.org-v1.yml
[INFO] Generating CRD keycloaks.k8s.keycloak.org:
[INFO]   - v1beta1 -> /home/bernhard/git/github/keycloak/keycloak/operator/target/classes/META-INF/fabric8/keycloaks.k8s.keycloak.org-v1beta1.yml
[INFO]   - v1 -> /home/bernhard/git/github/keycloak/keycloak/operator/target/classes/META-INF/fabric8/keycloaks.k8s.keycloak.org-v1.yml
[INFO] 
[INFO] --- resources:3.2.0:testResources (default-testResources) @ keycloak-operator ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Using 'UTF-8' encoding to copy filtered properties files.
[INFO] Copying 17 resources
[INFO] 
[INFO] --- compiler:3.8.1:testCompile (default-testCompile) @ keycloak-operator ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 20 source files to /home/bernhard/git/github/keycloak/keycloak/operator/target/test-classes
[WARNING] /home/bernhard/git/github/keycloak/keycloak/operator/src/test/java/org/keycloak/operator/testsuite/integration/BaseOperatorTest.java:[227,63] unchecked method invocation: method select in interface jakarta.enterprise.inject.Instance is applied to given types
  required: java.lang.Class<U>,java.lang.annotation.Annotation[]
  found:    java.lang.Class
[WARNING] /home/bernhard/git/github/keycloak/keycloak/operator/src/test/java/org/keycloak/operator/testsuite/integration/BaseOperatorTest.java:[227,64] unchecked conversion
  required: java.lang.Class<U>
  found:    java.lang.Class
[WARNING] /home/bernhard/git/github/keycloak/keycloak/operator/src/test/java/org/keycloak/operator/testsuite/integration/BaseOperatorTest.java:[228,90] unchecked method invocation: method select in interface jakarta.enterprise.inject.Instance is applied to given types
  required: java.lang.Class<U>,java.lang.annotation.Annotation[]
  found:    java.lang.Class
[WARNING] /home/bernhard/git/github/keycloak/keycloak/operator/src/test/java/org/keycloak/operator/testsuite/integration/BaseOperatorTest.java:[228,91] unchecked conversion
  required: java.lang.Class<U>
  found:    java.lang.Class
[WARNING] /home/bernhard/git/github/keycloak/keycloak/operator/src/test/java/org/keycloak/operator/testsuite/integration/BaseOperatorTest.java:[229,51] unchecked method invocation: method contextFor in class io.javaoperatorsdk.operator.api.config.Utils is applied to given types
  required: io.javaoperatorsdk.operator.api.config.ControllerConfiguration<?>,java.lang.Class<? extends io.javaoperatorsdk.operator.api.reconciler.dependent.DependentResource>,java.lang.Class<? extends java.lang.annotation.Annotation>
  found:    io.javaoperatorsdk.operator.api.config.ControllerConfiguration<capture#1 of ?>,java.lang.Class,java.lang.Class<io.javaoperatorsdk.operator.api.reconciler.dependent.Dependent>
[WARNING] /home/bernhard/git/github/keycloak/keycloak/operator/src/test/java/org/keycloak/operator/testsuite/integration/BaseOperatorTest.java:[229,67] unchecked conversion
  required: java.lang.Class<? extends io.javaoperatorsdk.operator.api.reconciler.dependent.DependentResource>
  found:    java.lang.Class
[WARNING] /home/bernhard/git/github/keycloak/keycloak/operator/src/test/java/org/keycloak/operator/testsuite/integration/ClusteringTest.java:[324,57] jsonMapper() in io.fabric8.kubernetes.client.utils.Serialization has been deprecated
[WARNING] /home/bernhard/git/github/keycloak/keycloak/operator/src/test/java/org/keycloak/operator/testsuite/integration/ClusteringTest.java:[335,48] jsonMapper() in io.fabric8.kubernetes.client.utils.Serialization has been deprecated
[WARNING] /home/bernhard/git/github/keycloak/keycloak/operator/src/test/java/org/keycloak/operator/testsuite/unit/KeycloakControllerTest.java:[47,53] unchecked conversion
  required: io.javaoperatorsdk.operator.api.reconciler.Context<org.keycloak.operator.crds.v2alpha1.deployment.Keycloak>
  found:    io.javaoperatorsdk.operator.api.reconciler.Context
[WARNING] /home/bernhard/git/github/keycloak/keycloak/operator/src/test/java/org/keycloak/operator/testsuite/unit/KeycloakControllerTest.java:[51,102] unchecked method invocation: method thenReturn in interface org.mockito.stubbing.OngoingStubbing is applied to given types
  required: T
  found:    io.fabric8.kubernetes.client.dsl.Resource
[WARNING] /home/bernhard/git/github/keycloak/keycloak/operator/src/test/java/org/keycloak/operator/testsuite/unit/KeycloakControllerTest.java:[51,103] unchecked conversion
  required: T
  found:    io.fabric8.kubernetes.client.dsl.Resource
[WARNING] /home/bernhard/git/github/keycloak/keycloak/operator/src/test/java/org/keycloak/operator/testsuite/unit/PodTemplateTest.java:[110,49] unchecked conversion
  required: io.javaoperatorsdk.operator.api.reconciler.Context<org.keycloak.operator.crds.v2alpha1.deployment.Keycloak>
  found:    io.javaoperatorsdk.operator.api.reconciler.Context
[INFO] 
[INFO] --- jar:3.1.2:jar (default-jar) @ keycloak-operator ---
[INFO] Building jar: /home/bernhard/git/github/keycloak/keycloak/operator/target/keycloak-operator-999.0.0-SNAPSHOT.jar
[INFO] 
[INFO] --- source:3.1.0:jar-no-fork (attach-sources) @ keycloak-operator ---
[INFO] Building jar: /home/bernhard/git/github/keycloak/keycloak/operator/target/keycloak-operator-999.0.0-SNAPSHOT-sources.jar
[INFO] 
[INFO] --- quarkus:3.13.3:build (default) @ keycloak-operator ---
[WARNING] [io.quarkus.config.build] The 'quarkus.operator-sdk.bundle.package-name' config property is deprecated and should not be used anymore
[INFO] [io.quarkiverse.operatorsdk.deployment.VersionAlignmentCheckingStep] QOSDK: 6.7.3 (commit: 2f026a2) built on Fri Aug 09 22:01:53 CEST 2024
[INFO] [io.quarkiverse.operatorsdk.deployment.VersionAlignmentCheckingStep] JOSDK: 4.9.2 (commit: fcfaaf1) built on Wed Jul 17 09:27:39 CEST 2024
[INFO] [io.quarkiverse.operatorsdk.deployment.OperatorSDKProcessor] Registered 'io.javaoperatorsdk.operator.processing.retry.GenericRetry' for reflection
[INFO] [io.quarkiverse.operatorsdk.deployment.OperatorSDKProcessor] Registered 'io.javaoperatorsdk.operator.processing.event.rate.LinearRateLimiter' for reflection
[INFO] [io.quarkiverse.operatorsdk.deployment.OperatorSDKProcessor] Registered 'io.quarkiverse.operatorsdk.runtime.QuarkusControllerConfiguration$DefaultRateLimiter' for reflection
[INFO] [io.quarkiverse.operatorsdk.deployment.OperatorSDKProcessor] Registered 'io.javaoperatorsdk.operator.processing.dependent.kubernetes.KubernetesDependentResourceConfig' for reflection
[INFO] [io.quarkiverse.operatorsdk.deployment.QuarkusControllerConfigurationBuildStep] Processed 'org.keycloak.operator.controllers.KeycloakRealmImportController' reconciler named 'keycloakrealmimportcontroller' for 'keycloakrealmimports.k8s.keycloak.org' resource (version 'k8s.keycloak.org/v2alpha1')
[INFO] [io.quarkiverse.operatorsdk.deployment.QuarkusControllerConfigurationBuildStep] Processed 'org.keycloak.operator.controllers.KeycloakController' reconciler named 'keycloakcontroller' for 'keycloaks.k8s.keycloak.org' resource (version 'k8s.keycloak.org/v2alpha1')
[INFO] Generating 'keycloakrealmimports.k8s.keycloak.org' version 'v2alpha1' with org.keycloak.operator.crds.v2alpha1.realmimport.KeycloakRealmImport (spec: org.keycloak.operator.crds.v2alpha1.realmimport.KeycloakRealmImportSpec / status org.keycloak.operator.crds.v2alpha1.realmimport.KeycloakRealmImportStatus)...
[WARNING] [io.quarkus.deployment.steps.ReflectiveHierarchyStep] Unable to properly register the hierarchy of the following classes for reflection as they are not in the Jandex index:
	- org.keycloak.representations.idm.RealmRepresentation (source: KubernetesClientProcessor > org.keycloak.operator.crds.v2alpha1.realmimport.KeycloakRealmImportSpec)
Consider adding them to the index either by creating a Jandex index for your dependency via the Maven plugin, an empty META-INF/beans.xml or quarkus.index-dependency properties.
[INFO] [io.quarkus.kubernetes.deployment.PropertyUtil] Kubernetes manifests are generated with 'The container port http' having default value '8080'. The app and manifests will get out of sync if the property 'quarkus.http.port' is changed at runtime.
Invalid AnsiLogger Stream -> Swapping to default sdt out logger.
[INFO] Adding existing Role with name: keycloak-operator-role.
[INFO] Adding existing ClusterRole with name: keycloak-operator-clusterrole.
[INFO] Adding existing RoleBinding with name: keycloak-operator-role-binding.
[INFO] Adding existing ClusterRoleBinding with name: keycloak-operator-clusterrole-binding.
[INFO] [io.quarkiverse.operatorsdk.deployment.AddRoleBindingsDecorator] Creating 'keycloakrealmimportcontroller-role-binding' RoleBinding to be applied to current namespace
[INFO] [io.quarkiverse.operatorsdk.deployment.AddRoleBindingsDecorator] Creating 'keycloakcontroller-role-binding' RoleBinding to be applied to current namespace
[INFO] Generating 'keycloaks.k8s.keycloak.org' version 'v2alpha1' with org.keycloak.operator.crds.v2alpha1.deployment.Keycloak (spec: org.keycloak.operator.crds.v2alpha1.deployment.KeycloakSpec / status org.keycloak.operator.crds.v2alpha1.deployment.KeycloakStatus)...
[INFO] [io.quarkiverse.operatorsdk.deployment.CRDGeneration] Generated keycloakrealmimports.k8s.keycloak.org CRD:
[INFO] [io.quarkiverse.operatorsdk.deployment.CRDGeneration]   - v1 -> /home/bernhard/git/github/keycloak/keycloak/operator/target/kubernetes/keycloakrealmimports.k8s.keycloak.org-v1.yml
[INFO] [io.quarkiverse.operatorsdk.deployment.CRDGeneration] Generated keycloaks.k8s.keycloak.org CRD:
[INFO] [io.quarkiverse.operatorsdk.deployment.CRDGeneration]   - v1 -> /home/bernhard/git/github/keycloak/keycloak/operator/target/kubernetes/keycloaks.k8s.keycloak.org-v1.yml
[INFO] [io.quarkiverse.operatorsdk.bundle.deployment.BundleProcessor] Assigning 'keycloakcontroller' reconciler to default bundle automatically named 'keycloak-operator'
[INFO] [io.quarkiverse.operatorsdk.bundle.deployment.BundleProcessor] Assigning 'keycloakrealmimportcontroller' reconciler to default bundle automatically named 'keycloak-operator'
[INFO] [io.quarkiverse.operatorsdk.bundle.deployment.BundleProcessor] Generating CSV for 'keycloak-operator' controller -> /home/bernhard/git/github/keycloak/keycloak/operator/target/bundle/keycloak-operator/manifests/keycloak-operator.clusterserviceversion.yaml
[INFO] [io.quarkiverse.operatorsdk.bundle.deployment.BundleProcessor] Generating annotations for 'keycloak-operator' controller -> /home/bernhard/git/github/keycloak/keycloak/operator/target/bundle/keycloak-operator/metadata/annotations.yaml
[INFO] [io.quarkiverse.operatorsdk.bundle.deployment.BundleProcessor] Generating bundle for 'keycloak-operator' controller -> /home/bernhard/git/github/keycloak/keycloak/operator/target/bundle/keycloak-operator/bundle.Dockerfile
[INFO] [io.quarkiverse.operatorsdk.bundle.deployment.BundleProcessor] Generating Custom Resource Definition for 'keycloak-operator' controller -> /home/bernhard/git/github/keycloak/keycloak/operator/target/bundle/keycloak-operator/manifests/keycloakrealmimports.k8s.keycloak.org-v1.crd.yml
[INFO] [io.quarkiverse.operatorsdk.bundle.deployment.BundleProcessor] Generating Custom Resource Definition for 'keycloak-operator' controller -> /home/bernhard/git/github/keycloak/keycloak/operator/target/bundle/keycloak-operator/manifests/keycloaks.k8s.keycloak.org-v1.crd.yml
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Quarkus augmentation completed in 5363ms
[INFO] 
[INFO] --- assembly:3.1.1:single (assemble-quarkus) @ keycloak-operator ---
[INFO] Reading assembly descriptor: assembly.xml
[INFO] Building zip: /home/bernhard/git/github/keycloak/keycloak/operator/target/keycloak-operator-999.0.0-SNAPSHOT.zip
[INFO] 
[INFO] --- assembly:3.1.1:single (assemble-bundle) @ keycloak-operator ---
[INFO] Reading assembly descriptor: olm-assembly.xml
[INFO] Building zip: /home/bernhard/git/github/keycloak/keycloak/operator/target/keycloak-operator-999.0.0-SNAPSHOT-olm.zip
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  20.517 s
[INFO] Finished at: 2024-09-01T17:52:56+02:00
[INFO] ------------------------------------------------------------------------
```

**Comment by baloo42** [comment]
@shawkins @andreaTP Is this a bug or a feature?

v1beta1 CRDs are only generated by the crd-generator-apt. Is v1beta1 still required?

**Comment by andreaTP** [comment]
I would say that V1beta is out of scope now

**Comment by shawkins** [comment]
Opened https://github.com/quarkusio/quarkus/issues/42991 so that we don't have to track a specific version of builder-annotations

**Comment by baloo42** [comment]
Just to be clear: 
I discovered this during some performance comparisons between quarkus-operator-sdk, crd-generator-apt and the new crd-generator maven plugin and couldn't find a reason why it is necessary to build the CRDs twice.

Was there a reason in the past?

It's not a bug which affects us but in my opinion Keycloak's operator is a best practice example for operators and as such it's maven build should follow best practices, too. In conclusion, anything which is not necessary, should be removed. Dependencies minimized etc...

**Comment by metacosm** [comment]
Using the Quarkus extension for the Java Operator SDK, you shouldn't indeed use the `crd-generator-apt` dependency since the Quarkus extension drives the generator directly from the API, while including that dependency will trigger the generation from the annotation processor.

**Comment by baloo42** [comment]
Thanks Chris, this confirms my assumptions. Then we can remove crd-generator-apt from the keycloak operator.

**Comment by andreaTP** [comment]
@baloo42 thanks for looking into this, historically I think we have been overriding the `crd-generator-apt` version, it might be not needed now.

I just want to settle expectations on this:

> in my opinion Keycloak's operator is a best practice example for operators and as such it's maven build should follow best practices, too

I love the Keycloak operator, it's maintained by excellent folks, and it's the Java Operator codebase I often refer too.
That said, being a "best practice example" has never been a goal for the project as far as I'm aware.

**Comment by baloo42** [comment]
> historically I think we have been overriding the crd-generator-apt version, it might be not needed now.

mhh, this sounds like there **was** at least a reason in the past. Do you think we have to further investigate, why it has been used?

> I love the Keycloak operator, it's maintained by excellent folks, and it's the Java Operator codebase I often refer too.
> That said, being a "best practice example" has never been a goal for the project as far as I'm aware.

I have no problem with reducing this statement to something like "the Keycloak operator has a broad visibility". ;)
But because of it's visibility there are often questions, why it is implemented as it is. (And here we are...)

Most of it is really self-explanatory and follows conventions (thanks!!!), but as I said, for this issue here I couldn't find a reason.

**Comment by andreaTP** [comment]
Fair :+1: 

Apparently it was me introducing it with this commit: https://github.com/keycloak/keycloak/commit/006aea300d7892fd60cf756535f25d2c325b3d1c

And, as it was a bulk change, it might have gone unnoticed to reviewers.

> Do you think we have to further investigate

If everything works as expected after removing it, I encourage you to proceed with the fix :+1:
